src/Tutor.Infrastructure/Database/EventStore/PostgreSQLEventStore/EventWrapper.cs

```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations.Schema;
+using Tutor.Core.BuildingBlocks.EventSourcing;
+
+namespace Tutor.Infrastructure.Database.EventStore.PostgreSqlEventStore
+{
+    internal class EventWrapper
+    {
+        public int Id { get; set; }
+        public int AggregateId { get; set; }
+        public DateTime Timestamp { get; set; }
+        public DomainEvent DomainEvent { get; set; }
+
+        public EventWrapper() { }
+
+        public EventWrapper(DomainEvent domainEvent)
+        {
+            DomainEvent = domainEvent;
+            AggregateId = domainEvent.AggregateId;
```

Comment:
I would actually really like to discuss improvements for this on our next sync.

As for why a wrapper currently exists: Events consist of general metadata common to all events (like the aggregate ID, time, usually aggregate version, event type) and flexible data specific to the event type. In the domain model, this is represented by extracting the common data to DomainEvent, and using inheritance to create concrete events with other data. I generally like the design of that, particularly because using it looks very expressive to me (and it's also, as far as I'm aware, the standard way to represent events). One modification I would like to make in that model is connecting event type names to subclasses of DomainEvent, but I'm not sure what the best way to do that is. 
However, when it comes to storage, we want to map the common data (the properties of DomainEvent) to normal columns, and the flexible data (the properties of subclasses of DomainEvent) to a jsonb column. It's not really possible to just tell EFCore and Npgsql to map a class hierarchy like that. As far as I'm aware, the only way to get the table we want with EFCore is to have an object that has properties which will be mapped to normal columns, and a property that will be mapped to jsonb. So what I did for now is exactly that: make a wrapper, add properties for the event metadata (aggregate ID and timestamp for now, I want to add event type as well), and a property for the flexible data. For the type of the flexible data property I just used DomainEvent, mostly because it was the simplest way to implement things right now.

One idea for a possibly better design that I just had: Generally speaking, the properties of DomainEvent are more related to event sourcing itself, while the properties of its subclasses are related to the actual domain. Maybe it would be better to actually just move the wrapper to BuildingBlocks - modify the DomainEvent class to basically just be an empty interface for the domain model to inherit from, and make a wrapper that contains the aggregate ID, timestamp, domain event, and maybe event type. Modify the Causes method in EventSourcedAggregate to receive a domain event, and then make a wrapper and add the metadata (which it should have access to anyway). This would be fairly simple to do, except that it would require making EventSourcedAggregate capable of accessing its ID (which I think should be done anyway).

Reply:
I think this idea has merit. We can go in that direction to see if it establishes a better overall design. Instead of Wrapper I might go for `EventSourcedDomainEvent` or something along those lines. Also, the timestamp might be suitable in the base `DomainEvent` as the rest of the information is not really useful without it.

Reply:
<code>DomainEvent</code> and <code>EventWrapper</code> were refactored in [59cafdb](https://github.com/Clean-CaDET/tutor/pull/19/commits/59cafdb3f5d9ed90f33d087d7d81c05e4c4eaaab).